### ISSUE TITLE
Wrong error message (or type system bug) for unsatisfied trait bound

### ISSUE THREAD

**Issue by defyrlt** [rust-lang/rust#29096]
Here's the [code](https://play.rust-lang.org/?gist=8295fe2b1cf5225ae082&version=nightly). Not sure how to get reduced example, sorry about that.
So, in general, the problem is that I have

``` rust
pub trait WhereType<'a>: ToSQL + CloneToTrait<'a> {}
```

and some implementations of it (32-37 lines). At 37 line I'm trying to implement this trait for struct which does not implement `ToSQL`. I would expect error message about that and instead I'm getting this:

``` rust
<anon>:37:1: 37:43 error: the trait `Numeric` is not implemented for the type `Subquery<'a>` [E0277]
<anon>:37 impl<'a> WhereType<'a> for Subquery<'a> {}
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<anon>:37:1: 37:43 help: see the detailed explanation for E0277
<anon>:37:1: 37:43 note: required by `WhereType`
<anon>:37:1: 37:43 error: the trait `core::fmt::Display` is not implemented for the type `Subquery<'a>` [E0277]
<anon>:37 impl<'a> WhereType<'a> for Subquery<'a> {}
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<anon>:37:1: 37:43 help: see the detailed explanation for E0277
<anon>:37:1: 37:43 note: `Subquery<'a>` cannot be formatted with the default formatter; try using `:?` instead if you are using a format string
<anon>:37:1: 37:43 note: required by `WhereType`
error: aborting due to 2 previous errors
playpen: application terminated with error code 101
```

It looks like rustc is trying to satisfy this impl instead: 

``` rust
impl<'a, T: Numeric + ToSQL + Clone> WhereType<'a> for T {}
```

Not sure if it's a bug, but it's definitely wrong/misleading error message.
It's the same for stable/beta/nightly.

**Comment by arielb1** [comment]
It picks the `impl<T:Numeric+Display> ToSql for T` impl.

**Comment by defyrlt** [comment]
Even worse because it's telling me that
```rust
<anon>:37:1: 37:43 note: required by `WhereType`
```
So it picks `impl<T: Numeric + Display> ToSql for T` for `Subquery` even though it's neither `Numeric` nor `Display`.

**Comment by estebank** [comment]
Current output:
```
error[E0309]: the parameter type `T` may not live long enough
  --> src/main.rs:33:38
   |
33 | impl<'a, T: Numeric + ToSQL + Clone> WhereType<'a> for T {}
   |          --                          ^^^^^^^^^^^^^
   |          |
   |          help: consider adding an explicit lifetime bound `T: 'a`...
   |
note: ...so that the type `T` will meet its required lifetime bounds
  --> src/main.rs:33:38
   |
33 | impl<'a, T: Numeric + ToSQL + Clone> WhereType<'a> for T {}
   |                                      ^^^^^^^^^^^^^

error[E0277]: the trait bound `&'a Subquery<'a>: Numeric` is not satisfied
  --> src/main.rs:37:10
   |
37 | impl<'a> WhereType<'a> for &'a Subquery<'a> {}
   |          ^^^^^^^^^^^^^ the trait `Numeric` is not implemented for `&'a Subquery<'a>`
   |
   = note: required because of the requirements on the impl of `ToSQL` for `&'a Subquery<'a>`

error[E0277]: `Subquery<'a>` doesn't implement `std::fmt::Display`
  --> src/main.rs:37:10
   |
37 | impl<'a> WhereType<'a> for &'a Subquery<'a> {}
   |          ^^^^^^^^^^^^^ `Subquery<'a>` cannot be formatted with the default formatter
   |
   = help: the trait `std::fmt::Display` is not implemented for `Subquery<'a>`
   = note: in format strings you may be able to use `{:?}` (or {:#?} for pretty-print) instead
   = note: required because of the requirements on the impl of `std::fmt::Display` for `&'a Subquery<'a>`
   = note: required because of the requirements on the impl of `ToSQL` for `&'a Subquery<'a>`
```

**Comment by estebank** [comment]
Update: the [current output](https://play.rust-lang.org/?gist=8295fe2b1cf5225ae082&version=nightly) points you in the right direction, but doesn't outright tell you to impl `ToSql` for `&Subquery`:

```
error[E0309]: the parameter type `T` may not live long enough
  --> src/main.rs:33:38
   |
33 | impl<'a, T: Numeric + ToSQL + Clone> WhereType<'a> for T {}
   |          --                          ^^^^^^^^^^^^^
   |          |
   |          help: consider adding an explicit lifetime bound...: `T: 'a +`
   |
note: ...so that the type `T` will meet its required lifetime bounds
  --> src/main.rs:33:38
   |
33 | impl<'a, T: Numeric + ToSQL + Clone> WhereType<'a> for T {}
   |                                      ^^^^^^^^^^^^^

error[E0277]: the trait bound `&'a Subquery<'a>: Numeric` is not satisfied
  --> src/main.rs:37:10
   |
31 | pub trait WhereType<'a>: ToSQL + CloneToTrait<'a> {}
   |                          ----- required by this bound in `WhereType`
...
37 | impl<'a> WhereType<'a> for &'a Subquery<'a> {}
   |          ^^^^^^^^^^^^^ the trait `Numeric` is not implemented for `&'a Subquery<'a>`
   |
   = note: required because of the requirements on the impl of `ToSQL` for `&'a Subquery<'a>`

error[E0277]: `Subquery<'a>` doesn't implement `std::fmt::Display`
  --> src/main.rs:37:10
   |
31 | pub trait WhereType<'a>: ToSQL + CloneToTrait<'a> {}
   |                          ----- required by this bound in `WhereType`
...
37 | impl<'a> WhereType<'a> for &'a Subquery<'a> {}
   |          ^^^^^^^^^^^^^ `Subquery<'a>` cannot be formatted with the default formatter
   |
   = help: the trait `std::fmt::Display` is not implemented for `Subquery<'a>`
   = note: in format strings you may be able to use `{:?}` (or {:#?} for pretty-print) instead
   = note: required because of the requirements on the impl of `std::fmt::Display` for `&'a Subquery<'a>`
   = note: required because of the requirements on the impl of `ToSQL` for `&'a Subquery<'a>`
```

**Comment by estebank** [comment]
Current output:

```
error[E0277]: the trait bound `&'a Subquery<'a>: Numeric` is not satisfied
  --> src/main.rs:37:28
   |
37 | impl<'a> WhereType<'a> for &'a Subquery<'a> {}
   |                            ^^^^^^^^^^^^^^^^ the trait `Numeric` is not implemented for `&'a Subquery<'a>`
   |
   = help: the following other types implement trait `Numeric`:
             i8
             i16
             i32
             i64
             u8
             u16
             u32
             u64
           and 2 others
note: required for `&'a Subquery<'a>` to implement `ToSQL`
  --> src/main.rs:25:28
   |
25 | impl<T: Numeric + Display> ToSQL for T {
   |         -------            ^^^^^     ^
   |         |
   |         unsatisfied trait bound introduced here
note: required by a bound in `WhereType`
  --> src/main.rs:31:26
   |
31 | pub trait WhereType<'a>: ToSQL + CloneToTrait<'a> {}
   |                          ^^^^^ required by this bound in `WhereType`

error[E0277]: `Subquery<'a>` doesn't implement `std::fmt::Display`
  --> src/main.rs:37:28
   |
37 | impl<'a> WhereType<'a> for &'a Subquery<'a> {}
   |                            ^^^^^^^^^^^^^^^^ `Subquery<'a>` cannot be formatted with the default formatter
   |
   = help: the trait `std::fmt::Display` is not implemented for `Subquery<'a>`
   = note: in format strings you may be able to use `{:?}` (or {:#?} for pretty-print) instead
   = help: the following other types implement trait `ToSQL`:
             Where<'a>
             &'a str
   = note: required for `&'a Subquery<'a>` to implement `std::fmt::Display`
note: required for `&'a Subquery<'a>` to implement `ToSQL`
  --> src/main.rs:25:28
   |
25 | impl<T: Numeric + Display> ToSQL for T {
   |                   -------  ^^^^^     ^
   |                   |
   |                   unsatisfied trait bound introduced here
note: required by a bound in `WhereType`
  --> src/main.rs:31:26
   |
31 | pub trait WhereType<'a>: ToSQL + CloneToTrait<'a> {}
   |                          ^^^^^ required by this bound in `WhereType`
```

**Comment by estebank** [comment]
[The current output (after fixing the lifetime errors)](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=4f77f7a5cc6b96ebc9abdfae45f56e6a) skips the last E0277 without many changes to the output since the last update:

```
error[E0277]: the trait bound `&'a Subquery<'a>: ToSQL` is not satisfied
  --> src/main.rs:37:28
   |
37 | impl<'a> WhereType<'a> for &'a Subquery<'a> {}
   |                            ^^^^^^^^^^^^^^^^ the trait `Numeric` is not implemented for `&'a Subquery<'a>`
   |
   = help: the following other types implement trait `Numeric`:
             f32
             f64
             i16
             i32
             i64
             i8
             u16
             u32
           and 2 others
note: required for `&'a Subquery<'a>` to implement `ToSQL`
  --> src/main.rs:25:28
   |
25 | impl<T: Numeric + Display> ToSQL for T {
   |         -------            ^^^^^     ^
   |         |
   |         unsatisfied trait bound introduced here
note: required by a bound in `WhereType`
  --> src/main.rs:31:26
   |
31 | pub trait WhereType<'a>: ToSQL + CloneToTrait<'a> {}
   |                          ^^^^^ required by this bound in `WhereType`
```

By pointing at line 25 and with the primary span label I think we already provide sufficient guidance on what is needed (`impl Numeric for &Subquery`).
